### ISSUE TITLE
TypeError: 'NoneType' object is not iterable

### ISSUE THREAD

**Issue by smhaider** [macarthur-lab/gnomad_browser#118]
after running python manage.py load_db it loaded the database for like 2 days and after loading I got this error. can you please let me know what would be the reason of it?

Done loading genes. Took 161925 seconds
Loaded 2400000 records from subset 2 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (161951 seconds)
Done indexing gene table. Took 52 seconds
Done loading transcripts. Took 20 seconds
Loaded 2500000 records from subset 0 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (162005 seconds)
Done indexing transcript table. Took 35 seconds
Loaded 2500000 records from subset 2 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (162113 seconds)
Done loading exons. Took 129 seconds
Loaded 2600000 records from subset 0 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (162203 seconds)
Loaded 2600000 records from subset 2 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (162441 seconds)
Loaded 2700000 records from subset 0 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (162582 seconds)
Done indexing exon table. Took 421 seconds
Started 0 processes to run load_gene_models
Dropped db.constraint.
Done loading constraint info. Took 90 seconds
Traceback (most recent call last):
  File "manage.py", line 80, in <module>
    manager.run()
  File "/home/syedhaider/exac/exac_env/local/lib/python2.7/site-packages/flask_script/__init__.py", line 366, in run
    raise e
TypeError: 'NoneType' object is not iterable
Loaded 2700000 records from subset 2 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (162700 seconds)
Finished loading subset 0 from  ExAC_HC.0.3.final.vep.vcf.gz (2793615 records)
Loaded 2800000 records from subset 2 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (162815 seconds)
Loaded 2900000 records from subset 2 of 4 from ExAC_HC.0.3.final.vep.vcf.gz (162914 seconds)
Finished loading subset 2 from  ExAC_HC.0.3.final.vep.vcf.gz (2998287 records)

**Comment by nawatts** [comment]
Hi @smhaider,

Based on that stack trace, it looks like you're running an [older revision of gnomad_browser](https://github.com/macarthur-lab/gnomad_browser/tree/5222151a674f7dcd72fface79d28acad7757701e).

I think the bug you're encountering is here:
https://github.com/macarthur-lab/gnomad_browser/blob/5222151a674f7dcd72fface79d28acad7757701e/exac.py#L395-L398

All the functions in this list
https://github.com/macarthur-lab/gnomad_browser/blob/5222151a674f7dcd72fface79d28acad7757701e/exac.py#L395
are expected to return a list of processes. However, at that revision, `load_constraint_information`, `load_cnv_models`, and `load_cnv_genes` do not.

This should be fixed in the [latest revision](aea8be7a52a8a07db0ec9337a8ecf8845e3c112c) of the code.

Alternatively, if you don't want to update, you can add `return []` to each of the `load_constraint_information`, `load_cnv_models`, and `load_cnv_genes` functions in exac.py

**Comment by smhaider** [comment]
Hello @nawatts 

Thank you so much for your response and detailed answers I have managed to fix this issue the browser is running but I have another problem now. during the loading process I encounter below mentioned errors, Although it does not stop the loading process but but still I am curious why I get these errors. please see the errors below. I have attached an image below ( so that you can have a better look at the traceback) kindly please have a look at it and let me know what could be the reason of it.

I look forward to hearing to hearing from you.
![whatsapp image 2018-05-28 at 11 21 56](https://user-images.githubusercontent.com/18593103/40717584-29ed5998-640d-11e8-91bc-b03e7902e01a.jpeg)

**Comment by niab** [comment]
Hi @smhaider,

This seems to be an old error and could be fixed by updating pymongo to the new version, see this issue: https://github.com/konradjk/exac_browser/issues/228

Also if you have a SSD drive (not external), I would advise you to use it to create the database as it will be much quicker (several hours). After that you can back up the database with _mongodump_ and restore it on a HDD drive (restoring from backup is much quicker than creating from VCF files).

**Comment by smhaider** [comment]
Hello @niab 

Thanks for your quick response and suggestions, just wanted to let you know that even after changing  "fields={"_id":False}" with "projection={"_id":False}" in exac.py and looksup.py it didn't work for me :/ can suggest something else?

Also I am using SSD 256 (not external) to create this database. 

Regards,
Haider

**Comment by niab** [comment]
Hi Haider,

Did you update pymongo and mongodb database too? Also see the second comment in the old issue where it was suggested to chunk the generator manually to solve this problem. Sorry for not precise answers, I was installing ExAC database quite a while ago and don't remember all the details (problems). 

Best Regards,

Nikita

**Comment by smhaider** [comment]
Dear Nikita,

Thanks again for your quick response. I have started loading again by making some more changes currently it is going good. 

I have one more question even after this error my browser runs but it shows less information as compared to current exac browser. do you think that this error is the reason behind that or I am missing something?

I am following the exact instructions provided here https://github.com/konradjk/exac_browser plus I am also using robo3t as I was facing some issues while assigning the db paths reason is still unknown. but it works fine. I just have less information as compared to current exac browser. can you please also highlight me what could be the reason for that? 

Kind Regards,
Haider

**Comment by niab** [comment]
Dear Haider,

Can you please specify what you mean by less? If there are less variants, than I would suggest to do the following:
1) Check that these variants are in your local database (you can use some GUI like MongoChef)
If there are in your database, but you don't see them in browser, than it is a problem with the browser.
2) If these variants are not in the database, than I would suggest to check if they are in the VCF files which you've used to create the database. If they are, than its a problem with import script. Otherwise you might be working with older version of the database.

Kind Regards,

Nikita

**Comment by niab** [comment]
Dear Haider,

Thanks for clarification, I also don't have CNVs and raw data in the local browser.
I think raw data (e.g. read depth on individual level) is not available at all.
CNVs and intolerance scores are available on FTP.
CNV: 
ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3.1/cnv
Intolerance Scores: 
ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3.1/functional_gene_constraint

For intolerance scores you'll have to rename the file or change a link to the file, see this row in exac.py file:

CONSTRAINT_FILE=os.path.join(os.path.dirname(__file__), EXAC_FILES_DIRECTORY, 'forweb_cleaned_exac_r03_march16_z_data_pLI_CNV-final.txt.gz')

For CNVs, I don't know which files have to be renamed or if some additional files are required which are not on the ftp. In exac.py they are named like this:
    CNV_FILE=os.path.join(os.path.dirname(__file__), EXAC_FILES_DIRECTORY, 'exac-gencode-exon.cnt.final.pop3'),
    CNV_GENE_FILE=os.path.join(os.path.dirname(__file__), EXAC_FILES_DIRECTORY, 'exac-final-cnvs.gene.rank'),

Kind Regards,

Nikita

**Comment by smhaider** [comment]
Dear Nikita,

Thanks again for your detailed answer. I have those files I will try with them. Thank you so much for all the answers, just wanted to tell you that loading is still in progress although I didn't get that error so far but it takes a lot of time. its been nearly 24 hours now. I was just wondering if it is normal that this thing takes time? 

Kind Regards,
Haider

**Comment by niab** [comment]
Dear Haider,

It depends... When I was trying to install ExAC database, the bottleneck was slow HDD. In the end I had to buy SSD drive as it just didn't work (I waited for several days). However, the database will work fine on HDD if you restore it from the dump (so I suggest you to make a dump for future after you install the database). You said that you work with SSD, so it shouldn't be a problem, but I would still suggest you to check few things:

1) Check how much memory is used. If it is full, than it could be a problem. Basically, it could be that the script is loading things into memory quicker than it writes them on the drive. In this case Operating System starts managing the memory by writing something to swap (which is some space on you drive) and that makes things even worse. I had 16GB of RAM, so in my case the problem was in HDD, not RAM.

2) I found the progress messages reported in console not very useful to understand what's going on. You can try to check how quickly it loads data to the database in GUI. Try to install MongoChef, connect to default port (localhost:27017, the name which you will have to input doesn't matter) and check how large is the database and how quickly it grows. See my screenshot for expected size for "variants" collection size. Note: you shouldn't have "exome_" and "genome_" collections as they are from gnomAD.

![mongochef](https://user-images.githubusercontent.com/8025563/40790418-c9dae8fa-64ec-11e8-92ac-ce392d031cf1.png)

After that, you can make a decision what to do. For example, if the database loads 10,000 variants per hour and you still need to load several million variants than there is probably no point to wait and you need to try something else.

By default the scripts try to load several things in parallel which could be a problem if hardware is not powerful enough to handle it, see this line in exac.py load_db function:

for load_function in [load_variants_file, load_dbsnp_file, load_base_coverage, load_gene_models, load_constraint_information, load_cnv_models, load_cnv_genes]:

You can try to load them one by one, just remove all the things from the list except 1 and see how it will work ( is the largest one).

- load_variants_file - is the main collection with variants data (the largest one)
- load_constraint_information, load_cnv_models, load_cnv_genes - this will probably fail to load unless you download the extra files and renamed them correctly.
- load_dbsnp_file, load_base_coverage, load_gene_models - these should load relatively quickly.

Hope it helps,

Nikita

**Comment by smhaider** [comment]
Dear Nikita,

that was indeed helpful. thank you so much for such a detailed reply and precious advises. I will go through them and get back to you. thanks again.

I have a couple of more questions regarding the input data itself (hope I can ask them here). will get back to you soon after getting this database run fine on my local machine.

Thanking you again for all your help and guidance.

Kind Regards,
Haider

**Comment by smhaider** [comment]
Dear Nikita,

I hope you are fine. Sorry to bother you again but I ran into another error :/ my db ran for like 12 days and now I got another error :/

![20180613_095513](https://user-images.githubusercontent.com/18593103/41338112-7b5eb0ee-6ef1-11e8-9a0a-e99b83ad7e1f.jpg)

I tried to search for this error couldn't find a solution to it. do you know what could be the reason of it?

I look forward to hearing from you.

Kind Regards,
Haider

**Comment by niab** [comment]
Dear Haider,

It seems that there is some problem  in constraint file (the one with pLI scores), as it can't parse data from it. If you want to understand what is wrong and fix it you'll need to check what's going on in "get_constraint_information" method in "parsing.py" file (e.g. try to add "print line" in "for line in constraint_file:" loop before "transcript, gene, chrom, info = line.strip().split(None, 3)" to check why it cannot be parsed). This collection is used only to display pLI scores and you can try to install it later. So you don't need to cancel everything just because of this error. However, 12 days is quite a lot, are you sure that it actually progressing well?

Kind Regards,

Nikita

**Comment by smhaider** [comment]
Dear Nikita, 

Yeah 12 days were quite allot. had memory issues actually that's why the loading process was quite slow. ok I will try to check what is wrong here thanks again for your help.

Kind Regards,
Haider

**Comment by smhaider** [comment]
@niab Dear Nikita,

I hope this email finds you well. I was really hoping to come up with a good news this time but unfortunately facing another issue and need your guidance again. kindly please have a look at the images below. now having problem with creating cache and running the database itself. 

![20180626_095659](https://user-images.githubusercontent.com/18593103/41897642-25dd5b98-7928-11e8-93a8-fde09429e147.jpg)
![20180626_095727_001](https://user-images.githubusercontent.com/18593103/41897647-2970894c-7928-11e8-8576-8e7589db3771.jpg)


Please have a look at the errors and let me know if you know how to fix it. I am trying to fix it from 2 days now but no success so far. :/

Kind Regards,
Haider

**Comment by niab** [comment]
Dear Haider,

I'm sorry, but I don't know how to fix them. I sometimes have errors "[Errno 32] Broken pipe", but it seems to be working fine. I also didn't manage to create cache, but as far as I understood it's just to have auto-complete when you type gene name in search field, so I can live without it.

Kind Regards,

Nikita

**Comment by smhaider** [comment]
Dear Nikita, 

Thank you for your quick response. oh ok no problem I would have ignored it too but the problem is even if I ignore creating cache when I try to run the database it fails to connect to the website and I am unable to retrieve an information although I can see the data in the database using Robo3t.
Anyhow thanks again.

@nawatts  I was just wondering maybe you can help me with this problem?

Kind Regards,
Haider

**Comment by smhaider** [comment]
@niab  Dear Nikita,

I hope you are fine. Finally a good news I have manage to solve all the errors and the Browser is working absolutely fine (except this IGV plot) thank you so much for all your assistance and taking time out of your hectic routine it really helped allot and I really appreciate that.

I have some general questions regarding the data itself. I wish to integrate my institute's data into it as I have very little to no knowledge of sequencing data so I have some questions regarding it I was just wondering if you can guide me in this regards as well?

I look forward to hearing from you.

Kind Regards,
Haider

**Comment by niab** [comment]
Dear Haider,

Good to hear, I would suggest you to create a backup using mongodump in case you'll have to install it again :). I don't think that there is an easy way to integrate custom data into ExAC (@nawatts please correct me if I am wrong). If you are not a skilled programmer I would suggest to do the opposite and add ExAC data to your institute data (e.g. write some scripts to get Allele Frequencies).

If you really want to do I can think of several approaches, but they all are not very good:

1) Check how variants data is stored in variants collection and update it with a script (e.g. change allele_count, hom_count, recalculate all frequencies etc). Pros: well, it's there... Cons: you'll "corrupt" original ExAC data and unless you add some extra flag to the database and modify ExAC browser to display you won't be able to distinguish your custom data from original)

2) Insert your variant data into variants collection using similar format. Pros: it's there and original data is not "corrupted" as you can write a script to delete all your variants. Cons: You might need to tweak your data (e.g. fake column values which you don't have); You'll see your data as duplicates on the website.

3) In gnomAD browser there were exome and genome variants were stored in 2 different collections and you can control what to display on the website. You can check gnomAD browser code and modify ExAC browser to do the same, but use your institution data instead of genomes. Pros: data is integrated, original data is not corrupted, no duplicates, can control what to display on the website. Cons: Will require a lot of programming, you will still need to tweak your data.

So, I would suggest not to integrate your data with ExAC. If you have some general questions about the data I can try to answer them, but I won't be able to help you with customizing ExAC browser.

Kind Regards,

Nikita

**Comment by smhaider** [comment]
@niab Dear Nikita,

Thanks again for your detailed answer. yeah I think it is better to add ExAC data. thanks again for your guidance. 

Kind Regards,
Haider

**Comment by smhaider** [comment]
@niab  Dear Nikita,

I hope you are doing well. I need your suggestion. I was just wondering if I use gnomAD exomes vcf and coverage files for ExAC browser is it a good idea or not? cause I am actually confused that If I replace exac data files with GnomAD exomes vcf and coverage files what I have to do with exac gencodes, MNP's and CNV's?? cause I can't find these files for in gnomAD bucket. 

Looking forward to hearing from you..

Kind Regards,
Haider

**Comment by niab** [comment]
Dear Haider,

I would suggest to set up gnomAD browser. You can use the same database for both projects (ExAC and gnomAD), so 1 database and 2 browsers will work fine. Have a look at my screen in this conversation, ExAC coverage collection is called "base_coverage", whereas gnomAD has "exome_coverage" and "genome_coverage" collections, you do not replace the data.

Kind Regards,

Nikita

**Comment by smhaider** [comment]
Dear Nikita,

Oh yes you are right I haven't noticed that thanks. ok then I will configure gnomAD too. Thanks a lot. 

Kind Regards,
Haider

**Comment by smhaider** [comment]
Dear Nikita,

Hope you are fine. just a quick question browser works absolutely fine except this small error that I get unable to figure out what it is. couldn't find the path as well as sqlite db. although website shows all the information and things seems to be alright. so I am just curious what could be the reason of this error? 

Error when accessing sqlite db: /mongo/readviz/combined_bams/1/combined_chr1_222.db - unable to open database file
Rendering variant: 1-55524222-C-T

Kind Regards,
HAider

**Comment by niab** [comment]
Dear Haider,

Sorry, I don't know the answer, I have this error too.

Kind Regards,

Nikita

**Comment by nawatts** [comment]
This repository is no longer being maintained. Development of the gnomAD browser has moved to [broadinstitute/gnomad-browser](https://github.com/broadinstitute/gnomad-browser).
